### PR TITLE
Refactoring monotone constraints (linked to #2305)

### PR DESCRIPTION
**Pull request by CharlesAuguste**
@aldanor @redditur

@guolinke @chivee @jameslamb  @huanzhang12 @Laurae2 @StrikerRUS 
You were all reviewers of the PR #2305. This PR was judged to be not merge-able because it is too big. Therefore, in my ultimate comment (https://github.com/microsoft/LightGBM/pull/2305#issuecomment-574560469) I said I would split into smaller PR easier to merge. This is the first PR in relation with #2305.

The goal of this PR is just to refactor the current code around monotone constraints. **The behavior of the library should not be affected by this PR**. Refactoring the monotone constraints that way was asked by @guolinke in the previous PR (from the comment https://github.com/microsoft/LightGBM/pull/2305#issuecomment-528675809). We think that refactoring is very important because it will make it easier to develop more efficient constraining methods. 

The main issue with this PR is that I was not able to refactor the files data_parallel_tree_learner.cpp, gpu_tree_learner.cpp, voting_parallel_tree_learner because I am not familiar with them (already mentioned in https://github.com/microsoft/LightGBM/pull/2305#issue-303471370 and in the discussion below). I can spend more time on it, but I would really appreciate getting help on that specific part if that is possible for you. @guolinke Since this refactor is something you were looking forward to, are you willing to fix these files for us? If you are familiar with these tree learners, I hope that would not be too much work. Please let me know what you think. Thank you very much.

### CODE REVIEW

File: src/treelearner/monotone_constraints.hpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#ifndef LIGHTGBM_TREELEARNER_MONOTONE_CONSTRAINTS_H_
+#define LIGHTGBM_TREELEARNER_MONOTONE_CONSTRAINTS_H_
+
+#include <algorithm>
```
Comment (aldanor) [inline src/treelearner/monotone_constraints.hpp]: @CharlesAuguste include vector and cstdint?

Comment (aldanor) [reply src/treelearner/monotone_constraints.hpp]: And limits

**Comment by guolinke** [comment]
As #2699 will significantly improve training efficiency.  Could you wait for its merge? it is almost ready, except the GPU codes.

**Comment by CharlesAuguste** [comment]
@guolinke I just rebased on your last PR. I also improved the compilation of this PR, and now only GPU is failing (but that's to be expected because I was not able to fix the file as explained in my initial comment). Let me know what you think. Thanks

**Comment by guolinke** [comment]
@CharlesAuguste I create PR in your repo

**Comment by CharlesAuguste** [comment]
@guolinke I merged your fix, we pass many more checks now. Thanks for that! However, there are still a few checks that we are not passing. These seem related to the docs and to the R package, but I don't understand how that is related to this PR. Any idea of what might be wrong? Thanks

**Comment by guolinke** [comment]
@CharlesAuguste I think they are temporary fails. 
Also, ping @StrikerRUS and @jameslamb .

**Comment by aldanor** [comment]
Yea, it looks like some docs-related server is down, causing temporary connection errors.

**Comment by StrikerRUS** [comment]
> Yea, it looks like some docs-related server is down, causing temporary connection errors.

I just re-run the test. Now everything is OK. Sorry about the inconvenience.

For R tests please refer to https://github.com/microsoft/LightGBM/pull/2748#issuecomment-583676147.

**Comment by jameslamb** [comment]
> > Yea, it looks like some docs-related server is down, causing temporary connection errors.
> 
> I just re-run the test. Now everything is OK. Sorry about the inconvenience.
> 
> For R tests please refer to [#2748 (comment)](https://github.com/microsoft/LightGBM/pull/2748#issuecomment-583676147).

@StrikerRUS @CharlesAuguste @aldanor I think #2752 will fix the R test issues, let's see.

**Comment by guolinke** [comment]
@CharlesAuguste could you fix the conflicts?

**Comment by CharlesAuguste** [comment]
No problem, I will do fix them.

**Comment by CharlesAuguste** [comment]
@guolinke I solved the merge conflicts, and all checks seem to pass. So it should be ready to merge if you are happy enough with the PR now :).

**Pull request merged**